File: linux-utils/hfunzip/hfunzip.c
#include <stdio.h>
#include <unistd.h>

int main(int argc, char *argv[]) {
  FILE *fptr;
  int count;
  char c;

  //check when no args are passed
  if(argc < 2) {
    fputs("hfunzip: file1 [file2 ...]\n",stdout);
    return 1;
  }

  //loop for every file passed
  for(int i = 1; (argv[i]); i++) {
    fptr = fopen(argv[i], "r");
    if(!fptr) {
      fputs("Error opening file",stdout);
      return 1;
    }
    
    //reads every 4 bytes and stores into count 
    while(fread(&count, 1, sizeof(int),fptr) == sizeof(int)) {
      //reads last 1 byte and stores into c
      if(fread(&c, 1, sizeof(char), fptr) != sizeof(char)) {
        return 1;
      }
      //prints chars based on count read
      while(count > 0) {
        printf("%c",c);
        count--;
      }
    }
    fclose(fptr);
  }
}

File: linux-utils/hfunzip/hfunzip.go
package main

import (
	"encoding/binary"
	"fmt"
	"log"
	"os"
)

func main() {
	argc := len(os.Args)
	//check if no args are passed
	if argc < 2 {
		os.Stderr.Write([]byte("hfunzip: file1 [file2 ...]\n"))
		os.Exit(1)
	}

	//create a buffer for count and char
	num := make([]byte,4) //first 4 bytes
	char := make([]byte, 1) //remaining 1 byte

	//unzip all files in order
	for _, arg := range os.Args[1:] {
		file, err := os.Open(arg)
		if err != nil {
			log.Fatal(err)
		}
		defer file.Close()

		//reading each num/char combo in file
		for {
			_, err := file.Read(num)
			if err != nil {
				break
			}
			_, err = file.Read(char)
			if err != nil {
				panic(err)
			}
			//convert binary count to int
			count := binary.LittleEndian.Uint32(num)
			var i uint32
			//print char based on count
			for i = 0; i < count; i++ {
				fmt.Printf("%c",char[0])
			}
		}
	}
}
